Create new dump_debug gtp command.

killerducky opened the pull request:
Only call it from ProductionJob when --debug dir is used.

This fixes https://github.com/gcp/leela-zero/issues/291

BTW I noticed dump_training is not listed in GTP list_commands. Is this on purpose to hide internal-use commands? Or should we update the list?

marcocalignano commented:
But why all the way down to game what should the object write in the debug. I would do the debug in the object Management and if the problem come from down we could bring it up with an Result of type ERROR

marcocalignano commented:
I do not see any gtp command called dump_debug

marcocalignano commented:
Oh, I see, **you** did put the new GTP command! Ok but then you could add a command line option for leela with the filename like `--debug debug.txt` and the dump debug knows where to dump the debug or even better you do like the two function `dumpTraining` and `writeSgf`: use the random name with a different extension.
All my suggestion are a way to avoid to change all those objects only to provide a file name, the otion for autogtp is good and this option would help a lot later when we are out of the test phase. but after that we can use the same method we use for save sgf and dump training (you would have also the same name for all thrre files)

marcocalignano commented:
Ok so maybe you just need a Boolean value debug true or false in` Job::init() ` you do not use the QString anyway and if you want to move out of the way in a sub directory you can do it in Management (like we copy the sgf if one wants to keep them) and you can pass this Boolean from `Management` to `Job` thru the `Work::order()` function.

killerducky commented:
Good suggestion. I will update my code to do something like that. Also I think the new validation/production code is still in flux and being debugged. This pull request is lower priority so I think we can wait to merge this until things stabilize more.

killerducky commented:
@marcocalignano I updated it to use Job::init() as you suggested, it is much cleaner now! I also updated the resign_analysis.py script to handle the case were some games have -r 0 and others have resign enabled.

marcocalignano commented:
@killerducky Looks good. You did a refactoring and you deleted the unused parameter. So I wait with my pull request till @gtp merge this one.

oni-link commented on autogtp/Game.cpp:
```diff
@@ -317,6 +317,16 @@ bool Game::dumpTraining() {
+    QTextStream(stdout) << "Dumping " << m_fileName + ".debug.txt" << endl;
+
+    if (!sendGtpCommand(qPrintable("dump_debug " +
+                        m_fileName + ".debug.txt"))) {
```
Did you consider replacing this with `return sendGtpCommand(...);`?

killerducky replied:
I kept the same style as the other functions around it. I leave it to the original author @marcocalignano to decide if he wants to change the style.

marcocalignano replied:
@killerducky  Change it to return please.

gcp replied:
I'll pull this next then.

killerducky replied:
fixed.
